Rendering/Effects/OffGeometryEffect.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright (c) 2021 Technical University of Munich
+Chair of Computational Modeling and Simulation.
+
+TUM Open Infra Platform is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License Version 3
+as published by the Free Software Foundation.
+
+TUM Open Infra Platform is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "OffGeometryEffect.h"
+
+
+#include "../Resources/RenderResources.h"
+#include <BlueFramework/Rasterizer/vertex.h>
+#include <BlueFramework/Engine/Mesh/geometryGeneration.h>
+
+#include <tuple>
+
+
+OIP_NAMESPACE_OPENINFRAPLATFORM_RENDERING_BEGIN
+
+
+OffGeometryEffect::OffGeometryEffect(buw::IRenderSystem * renderSystem, buw::ReferenceCounted<buw::IViewport> viewport, buw::ReferenceCounted<buw::ITexture2D> depthStencilMSAA, buw::ReferenceCounted<buw::IConstantBuffer> worldBuffer)
+	: EffectBase(renderSystem), viewport_(viewport), worldBuffer_(worldBuffer), depthStencilMSAA_(depthStencilMSAA)
+{
+}
+
+OffGeometryEffect::~OffGeometryEffect()
+{
+	worldBuffer_ = nullptr;
+	vertexBuffer_ = nullptr;
+	indexBuffer_ = nullptr;
+	depthStencilMSAA_ = nullptr;
+	pipelineState_ = nullptr;
+	viewport_ = nullptr;
+}
+
+void OffGeometryEffect::setOffModel(buw::ReferenceCounted<Core::OffConverter::OffModel> offModel)
+{
+	offModel_ = offModel;
+}
+
+void OffGeometryEffect::changeOffset(const buw::Vector3d& offsetOld, const buw::Vector3d& offsetNew)
+{
+	if (!offModel_)
+		return;
+
+	if(!offModel_->isEmpty()) {
+		
+		//reset 
+		if(vertexBuffer_) vertexBuffer_ = nullptr;
+		if(indexBuffer_) indexBuffer_ = nullptr;
+
+		//tmp containers
+		std::vector<buw::Vector3d> verticesMesh;
+		std::vector<uint32_t> indicesMesh;
+
+		//process the data
+		if (!offModel_->geometry().isEmpty())
+		{
+			std::vector<buw::Vector3d> vertices;
+			vertices.reserve(offModel_->geometry().vertices.size());
+			for (auto& vertex : offModel_->geometry().vertices)
+			{
+				buw::Vector3d vtx(vertex);
+				vtx[0] += offsetNew.x();
+				vtx[1] += offsetNew.y();
+				vtx[2] += offsetNew.z();
+				vertices.push_back(vtx);
+			}
+
+			const uint32_t verticesOffset = verticesMesh.size();
```

Comment:
Is always 0 since the vector is initialized empty. Please replace with 0 for clarity or remove.

Reply:
Resolved: b999700